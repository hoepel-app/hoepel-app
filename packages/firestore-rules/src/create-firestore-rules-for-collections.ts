import { Rule, Collection } from '@hoepel.app/schema-firestore';

export const createFirestoreRulesForCollections = (collections: readonly Collection[], extraRules: readonly Rule[] = []): string => {
    const rulesFromSchemaAsText = collections
        .map(collection => collection.firestoreRules)
        .map(rules => rules.map(rule => rule.toString()).join('\n\n'))
        .join('\n\n');

    const extraRulesText = extraRules.map(rule => rule.toString()).join('\n\n')

    return [
        'service cloud.firestore {',
        indent('/// Extra rules provided for generation'),
        indent(extraRulesText),
        indent('/// Automaticaly generated Firestore rules'),
        indent(rulesFromSchemaAsText),
        '}'
    ].join('\n\n')
}

const indent = (str: string, indentation = '  '): string => str
    .split('\n')
    .map(line => indentation + line)
    .join('\n')
