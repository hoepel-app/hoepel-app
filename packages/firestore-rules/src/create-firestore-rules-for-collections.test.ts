import {
  Rule,
  IndexedByTenantCollection,
  DocPerTenantCollection,
  DocPerUserCollection,
} from '@hoepel.app/schema-firestore'
import { createFirestoreRulesForCollections } from './create-firestore-rules-for-collections'

describe('createFirestoreRulesForCollections', () => {
  it('creates Firestore rules for a list of collections', () => {
    const created = createFirestoreRulesForCollections(
      [
        new IndexedByTenantCollection('tenant-indexed', {
          writePermission: 'write-permission',
          readPermission: 'read-permission',
          updatePermission: 'update-permission',
          deletePermission: 'delete-permission',
        }),
        new DocPerTenantCollection('doc-per-tenant', {
          readPermission: 'doc-read-perm',
          writePermission: 'doc-write-perm',
        }),
        new DocPerUserCollection('doc-per-user'),
      ],
      [
        new Rule({
          match: '/my/match/thingie',
          comment: 'My extra comment',
          allowCreate: true,
          allowRead: 'my custom thingie',
          allowUpdate: false,
          allowDelete: true,
        }),
      ]
    )

    expect(created).toMatchSnapshot()
  })
})
