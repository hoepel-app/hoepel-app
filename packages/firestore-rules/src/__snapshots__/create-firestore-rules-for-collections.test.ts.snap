// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFirestoreRulesForCollections creates Firestore rules for a list of collections 1`] = `
"service cloud.firestore {

  /// Extra rules provided for generation

  // My extra comment
  match /my/match/thingie {
    allow create: if true;
    allow read  : if my custom thingie;
    allow update: if false;
    allow delete: if true;
  }

  /// Automaticaly generated Firestore rules

  // Generated permissions for collection 'tenant-indexed' (collection contains docs with doc.tenant == tenant name)
  match /databases/{database}/documents/tenant-indexed/{document} {
    function hasPermission(permission, tenant) {
      return permission in get(/databases/$(database)/documents/users/$(request.auth.uid)/tenants/$(tenant)).data.permissions;
    }
    allow create: if hasPermission('write-permission', request.resource.data.tenant);
    allow read  : if hasPermission('read-permission', resource.data.tenant);
    allow update: if hasPermission('update-permission', resource.data.tenant);
    allow delete: if hasPermission('delete-permission', resource.data.tenant);
  }
  
  // Generated permissions for collection 'doc-per-tenant' (collection contains docs with doc.id == tenant name)
  match /databases/{database}/documents/doc-per-tenant/{document} {
    function hasPermission(permission, tenant) {
      return permission in get(/databases/$(database)/documents/users/$(request.auth.uid)/tenants/$(tenant)).data.permissions;
    }
    allow create: if hasPermission('doc-write-perm', document);
    allow read  : if !exists(/databases/$(database)/documents/doc-per-tenant/$(document)) || hasPermission('doc-read-perm', document);
    allow update: if hasPermission('doc-write-perm', document);
    allow delete: if false;
  }
  
  // Generated permissions for collection 'doc-per-user' (collection contains docs with doc.id == user id)
  match /databases/{database}/documents/users/{uid} {
    allow create: if false;
    allow read  : if uid == request.auth.uid;
    allow update: if false;
    allow delete: if false;
  }
  
  // Generated permissions for collection 'doc-per-user' (collection contains docs with doc.id == user id)
  match /databases/{database}/documents/users/{uid}/{document=**} {
    allow create: if false;
    allow read  : if uid == request.auth.uid;
    allow update: if false;
    allow delete: if false;
  }
  
  // Generated permissions for collection 'doc-per-user' (collection contains docs with doc.id == user id) - Admin access
  match /databases/{database}/documents/users/{document=**} {
    allow create: if false;
    allow read  : if request.auth.token.isAdmin;
    allow update: if request.auth.token.isAdmin;
    allow delete: if false;
  }

}"
`;
